decisions/lock checing.md

```diff
@@ -0,0 +1,46 @@
+# XXXX. Use static Thread Safety Analysis annotations
+
+Date: 2022-03-21
+
+## Status
+
+Proposed
+
+## Context
+
+Some mistakes in lock usage can be detected statically.
+As opposed to dynamic approaches, either TSan (already in use) or lock priorities (proposed in [2], not implemented), the static approach proposed now can be checked reliably in the compiler, or in the IDE (if based on clangd).
+
+The way it works is that functions are annotated with what locks they require to be held, and the compiler then checks these conditions are satisfied.
+
+Some additional features of the Clang solution, namely, the TA_GUARDED() annotation for struct members, are available only in C++ [3]; meaning these features may not be used in skupper-router. 
+
+[1] https://clang.llvm.org/docs/ThreadSafetyAnalysis.html
+[2] 
+
+## Decision
+
+Annotate all functions that acquire, release, or require locks with the Clang Thread Safety Analysis annotations.
+Run a CI job that will check these annotations.
+
+## Consequences
+
+Header file defining the annotations needs to be included.
+There are multiple possibilities, either the Clang one(mutex.h), or the Zircon one (thread_annotations.h).
```

Comment:
This one is licensed under Apache, https://github.com/google/google-input-tools/blob/master/client/base/thread_annotations.h